### ISSUE TITLE
Loading weights/cfg from memory

### ISSUE THREAD

**Issue by Kaimbelic** [AlexeyAB/darknet#6019]
I have an application that requires the weights to be encrypted on disk. As such the decrypted weights can not be saved to disk and loaded. Looking for a constructor that could load the decrypted weights stored inside of std::string .

The current constructor for Detector requires the file to be on disk.
`Detector::Detector(std::string cfg_filename, std::string weight_filename, int gpu_id)`

The feature request would be to have a constructor option that allows for the input strings to be the actual weight file from memory.

`Detector::Detector(std::string cfg_filename, std::string weight_mem, int gpu_id)`

Thank you for your time.

**Comment by grandec** [comment]
I think there should be an option that both cfg and weights to be loaded from memory, however I dont know how the constructor signature should work because in both cases the method signature are the same.

**Comment by Kaimbelic** [comment]
I am not too sure for the signature either. Clearly they would have the same signature. Could add an extra flag for if the data is from disk or from memory? Not sure if that is a good way either.

**Comment by grandec** [comment]
The flag might be a good solution, however the names of the parameters will be misleading, because the user will assume that is a filepath and not a filebuffer.

**Comment by Kaimbelic** [comment]
Yeah not an easy solution. Would need to find a better naming convention for the inputs and create the flag such that it states the mode at which the inputs will be interpreted.

**Comment by imaami** [comment]
I don't understand the focus on repurposing that particular constructor, or why adding a flag parameter but keeping the `std::string` type would make sense. The normal way to extend the API would be to simply add an overloaded constructor like so:

`Detector::Detector(std::string cfg_filename, std::uint8_t const *weight_data, std::size_t weight_data_size, int gpu_id)`

I can't think of a situation where a binary blob such as a weights data file should ever be a `std::string`. It would probably not work at all. What you need is a byte buffer and its size.

**Comment by grandec** [comment]
> I don't understand the focus on repurposing that particular constructor, or why adding a flag parameter but keeping the `std::string` type would make sense. The normal way to extend the API would be to simply add an overloaded constructor like so:
> 
> `Detector::Detector(std::string cfg_filename, std::uint8_t const *weight_data, std::size_t weight_data_size, int gpu_id)`
> 
> I can't think of a situation where a binary blob such as a weights data file should ever be a `std::string`. It would probably not work at all. What you need is a byte buffer and its size.

Oh, that's right. I completely forgot about the weight_data_size. With this constructor that you showed, a simple overload would do the trick.

**Comment by stephanecharette** [comment]
If you guys are going down this path, then make it C++, not a hybrid between C and C++.  I recommend:

    Detector::Detector(const std::string & cfg_filename, const std::vector<uint8_t> & weights, int gpu_id);

**Comment by imaami** [comment]
> If you guys are going down this path, then make it C++, not a hybrid between C and C++. I recommend:
> 
> ```
> Detector::Detector(const std::string & cfg_filename, const std::vector<uint8_t> & weights, int gpu_id);
> ```

Why use `std::vector` for a byte buffer? Just to have the size parameter inside it (instead of a separate function argument?

Is it guaranteed that a `std::vector<uint8_t>` contains an unpadded byte array? And can you construct that vector without literally copying the entire buffer (just for the purpose of having a vector type as the function argument)?

**Comment by imaami** [comment]
@stephanecharette continuing from my previous message:

After refreshing my memory on [`std::vector`](https://en.cppreference.com/w/cpp/container/vector) I understand you're right about `std::vector<uint8_t>` being an appropriate type. :) It's sequential and provides a pointer to the underlying buffer without copying it.

One problem still stands: I don't think you can get a `std::vector<uint8_t>` from a `uint8_t *` and `size_t` without making a deep copy. Therefore adding only the `std::vector<uint8_t>` constructor would force those starting with a plain byte buffer to do a pointless vector allocation and copy for potentially quite a lot of data.

So with that in mind, what about this?

`Detector::Detector(std::string const &cfg_filename, std::uint8_t const *weight_data, std::size_t weight_data_size, int gpu_id);

Detector::Detector(std::string const &cfg_filename, std::vector<std::uint8_t> const &weights, int gpu_id)
    : Detector(cfg_filename, weights.data(), weights.size(), gpu_id)
{
}`

**Comment by grandec** [comment]
> @stephanecharette continuing from my previous message:
> 
> After refreshing my memory on [`std::vector`](https://en.cppreference.com/w/cpp/container/vector) I understand you're right about `std::vector<uint8_t>` being an appropriate type. :) It's sequential and provides a pointer to the underlying buffer without copying it.
> 
> One problem still stands: I don't think you can get a `std::vector<uint8_t>` from a `uint8_t *` and `size_t` without making a deep copy. Therefore adding only the `std::vector<uint8_t>` constructor would force those starting with a plain byte buffer to do a pointless vector allocation and copy for potentially quite a lot of data.
> 
> So with that in mind, what about this?
> 
> ```
> Detector::Detector(std::string const &cfg_filename, std::uint8_t const *weights_data,
>                    std::size_t weights_data_size, int gpu_id);
> 
> Detector::Detector(std::string const &cfg_filename, std::vector<std::uint8_t> const &weights, int gpu_id)
>     : Detector(cfg_filename, weights.data(), weights.size(), gpu_id) { }
> ```

I have a doubt, if you know that you are agoing to use need vector, why not put the data directly in it? with the reserve method and so on?

**Comment by imaami** [comment]
> > @stephanecharette continuing from my previous message:
> I have a doubt, if you know that you are agoing to use need vector, why not put the data directly in it? with the reserve method and so on?

Why force the use of std::vector when the input data is not inherently a C++ vector but rather just a binary data blob?

What if the user _can't_ decide to use a vector from the start, but instead receives the data from some other library as a normal C-style byte buffer? In that case it would make no sense to require that everyone should use a vector.

Also, the only thing that the constructor would do with the vector is literally just get the buffer pointer and size, as seen in your current PR:

https://github.com/grandec/darknet/blob/577f78604f6a4a616c21d2313ec3a53b7af86848/src/yolo_v2_class.cpp#L201-L204

It would make sense for the new constructor to use `std::uint8_t const *, std::size_t` as the weights input arguments, and then add an inline wrapper constructor that accepts `std::vector<std::uint8_t> const &` and calls the former. That way no one is forced to do useless copying of memory, but we also get a vector-based constructor for free. Win-win.

**Comment by imaami** [comment]
@Kaimbelic Just to confirm: you only need the weights data to be in memory, not the cfg file text as well? Do you still read the cfg normally as a text file from disk?

**Comment by grandec** [comment]
> > > @stephanecharette continuing from my previous message:
> > > I have a doubt, if you know that you are agoing to use need vector, why not put the data directly in it? with the reserve method and so on?
> 
> Why force the use of `std::vector` when the input data is not inherently a C++ vector but rather just a binary data blob? Shouldn't the default course of action be to _first_ add support for the most widely-used and simple type?
> 
> What if the user _can't_ decide to use a vector from the start, but instead receives the data from some other library as a normal C-style byte buffer? In that case it would make no sense to require that everyone should use a vector.
> 
> Also, the only thing that the constructor would do with the vector is literally just get the buffer pointer and size, as seen in your current PR:
> 
> https://github.com/grandec/darknet/blob/577f78604f6a4a616c21d2313ec3a53b7af86848/src/yolo_v2_class.cpp#L201-L204
> 
> It would make sense for the new constructor to use `std::uint8_t const *, std::size_t` as the weights input arguments, and then add an inline wrapper constructor that accepts `std::vector<std::uint8_t> const &` and calls the former. That way no one is forced to do useless copying of memory, but we also get a vector-based constructor for free. Win-win.

I think you are right, it completely makes sense.

**Comment by grandec** [comment]
I pushed another commit at #6068 with @imaami latest suggestions.

**Comment by Kaimbelic** [comment]
> 
> 
> @Kaimbelic Just to confirm: you only need the weights data to be in memory, not the cfg text as well? Do you still read the cfg normally as a text file from disk?

Yes, just the weights.
